cmd/kubectl-gadget/utils/trace.go

```diff
@@ -208,65 +195,243 @@ func GenericTraceCommand(
+	}
+
+	for _, trace := range traces.Items {
+		localError := updateTraceOperation(&trace, operation)
+		if localError != nil {
+			err = fmt.Errorf("%w\nError updating trace operation for %s: %v", err, traceID, localError)
+		}
+	}
+
+	return err
+}
+
+// PrintTraceOutputFromStream is used to print trace output using generic
+// printing function.
+// This function is likely to be used by trace which has TraceOutputMode set to
+// Stream.
+func PrintTraceOutputFromStream(traceID string, config *TraceConfig, transformLine func(string) string) error {
+	return printTraceOutput(traceID, config, nil, transformLine)
+}
+
+// PrintTraceOutputFromStatus is used to print trace output using function
+// pointer provided by caller.
+// It will parse trace.Spec.Output and print it calling the function pointer.
+func PrintTraceOutputFromStatus(traceID string, customResultsDisplay func(nodes *corev1.NodeList, results *gadgetv1alpha1.TraceList)) error {
+	return printTraceOutput(traceID, nil, customResultsDisplay, nil)
+}
+
+// printTraceOutput is used to print trace output.
+// Trace state must be either "Started" or "Completed" to contain some events.
+func printTraceOutput(traceID string, config *TraceConfig,
+	customResultsDisplay func(nodes *corev1.NodeList, results *gadgetv1alpha1.TraceList),
+	transformLine func(string) string) error {
 	// TODO: use a watcher to avoid looping client-side
-	//
-	// watch, err := traceRestClient.
-	//	Get().
-	//	Namespace("gadget").
-	//	Resource("traces").
-	//	VersionedParams(&listTracesOptions, scheme.ParameterCodec).
-	//	Watch(context.TODO())
-	// if err != nil {
-	//	contextLogger.Fatalf("Error waiting for traces: %q", err)
-	// }
-	// for event := range watch.ResultChan() {
-	//	fmt.Printf("Event: %v\n", event)
-	//	if data, ok := event.Object.(*metav1.Status); ok {
-	//		contextLogger.Infof("watcher status: %s", data.Message)
-	//	} else if t, ok := event.Object.(*gadgetv1alpha1.Trace); ok {
-	//		fmt.Printf("Got: %v\n", t)
-	//	} else {
-	//		contextLogger.Fatalf("Error waiting for traces: got unexpected %v", event)
-	//	}
-	// }
-
-	var results gadgetv1alpha1.TraceList
+	var err error
+	var trace *gadgetv1alpha1.Trace
+	var traces gadgetv1alpha1.TraceList
+
 	start := time.Now()
 RetryLoop:
 	for {
-		results = gadgetv1alpha1.TraceList{}
-		err = traceRestClient.
-			Get().
-			Namespace("gadget").
-			Resource("traces").
-			VersionedParams(&listTracesOptions, scheme.ParameterCodec).
-			Do(context.TODO()).
-			Into(&results)
+		traces, err = getTraceListFromID(traceID)
 		if err != nil {
-			deleteTraces(contextLogger, traceRestClient, traceID)
-			contextLogger.Fatalf("Error getting traces: %q", err)
+			return err
 		}
 
 		timeout := time.Since(start) > traceTimeout
 		successNodeCount := 0
 		nodeErrors := make(map[string]string)
 		nodeWarnings := make(map[string]string)
-		for _, i := range results.Items {
+		for _, i := range traces.Items {
+			trace = &i
 			if i.Status.OperationError != "" {
```

Comment:
I am implementing the biolatency gadget on top of this PR and I found that this `for` loop needs to be modified because it will not work for the multi-rounds gadgets. 

I think it is because now each gadget needs to specify what is the `Status.State` it is waiting for. Without the multi-rounds gadgets, we avoided that need by adding an [internal logic](https://github.com/kinvolk/inspektor-gadget/blob/353f9a2a91211a9032cfd142cb21b884d8009b71/cmd/kubectl-gadget/utils/trace.go#L439) but now it is more difficult. Consider the following situation:

We create a trace and set the `start` operation:
```
$ ./kubectl-gadget biolatency start --node worker
P0HCLetshkIzyMzI
```

If the operation successes, the `Status.State` will go in `Started`. Then, if we execute `./kubectl-gadget biolatency stop P0HCLetshkIzyMzI`, this `for` loop will consider that the output is ready if the `Status.State` is `Completed` **or** `Started`. And this is the problem because our biolatency trace is already in `Stated` so this loop will immediately go out and it will not actually wait for the output (which will be ready when the `Status.State` moves to `Completed`). 

One solution is that the caller must specify which `Status.State` it is waiting for. For instance, in my previous example, it will be `Completed`, while the stream gadgets will always wait for `Started`. However, I think that the function `printTraceOutput` is already difficult to follow because it contains the logic for both output modes (Stream and Status) thus I don't want to add more complexity. 

Instead, I propose to move each output mode logic to their respective callers `PrintTraceOutputFromStream` and `PrintTraceOutputFromStatus` and the common function `printTraceOutput` could be renamed as `waitForTraceUpdate` which, among others, receives the `Status.State` it must wait for. In the end, this common function will contain mainly this `for` loop ([link](https://github.com/kinvolk/inspektor-gadget/blob/353f9a2a91211a9032cfd142cb21b884d8009b71/cmd/kubectl-gadget/utils/trace.go#L416-L478)) and then each caller will call the `genericStreamsDisplay` or the `customResultsDisplay`.

What do you think?